### PR TITLE
0.10.0

### PR DESCRIPTION
**Pull request by BossZou**

### CODE REVIEW

File: ann_benchmarks/runner.py [hunk 1]
```diff
@@ -223,6 +232,8 @@ def run_docker(definition, dataset, count, runs, timeout, batch, cpu_limit,
         cpu_limit = "0-%d" % (multiprocessing.cpu_count() - 1)
     print('Running on CPUs:', cpu_limit)
 
+    logic_cpu_num = psutil.cpu_count(logical=True)
```
Comment (erikbern) [inline ann_benchmarks/runner.py:235]: We've been running the benchmark in a single-cpu mode for quite some time. The argument is that most algorithms are single-threaded, but can be trivially parallelizeable by just using multiple threads for the queries anyway, so it doesn't really affect the throughput per CPU anyway.

**Comment by erikbern** [comment]
This looks great. I'm fine merging this (although I would like to preserve the CPU logic the way it is, not the proposed change). I'm currently running a benchmark suite, and a few Milvus runs already completed, but I can wipe those so they use the new code.

File: templates/summary.html [hunk 2]
```diff
@@ -1,60 +1,68 @@
...
+            >
+          </div>
+        </div>
+        {% endfor %}
+      </div>
+      {% endfor %}
+    </div>
+  </div>
+  {% endif %} {% endfor %} {% endblock %}
+</div>
```
[the remark this answers is not in the document]
Comment (maumueller) [reply templates/summary.html:68]: Additionally, there is no visible difference between batched and non-batched queries any longer. 

image

(bottom one is the batched run from ann-benchmarks.com)

the labels are also a bit odd in the quicklinks section (maybe difficult to change though with your approach)

image

File: ann_benchmarks/algorithms/milvus_hnsw.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        if self._metric == 'angular':
+            v /= numpy.linalg.norm(v)
+        v = v.tolist()
+        search_param = {
+            "ef": self._ef
+        }
+        future = self._milvus.search(collection_name=self._table_name, query_records=[v], top_k=n, params=search_param, _async=True)
+        return future
+
+    def handle_query_list_result(self, query_list):
```
Comment (erikbern) [inline ann_benchmarks/algorithms/milvus_hnsw.py:72]: why is this method needed? can't it just be handled through the `query` method? we already time the calls in the runner

File: ann_benchmarks/runner.py [hunk 4]
```diff
@@ -71,12 +64,28 @@ def batch_query(X):
...
+            return (total, candidates)
+
         if batch:
             results = batch_query(X_test)
+            handle_time = 0
         else:
-            results = [single_query(x) for x in X_test]
+            query_list = [single_query(x) for x in X_test]
+            handle_time, handled_list = algo.handle_query_list_result(query_list)
+            results = [get_candidates(l) for l in handled_list]
```
[the remark this answers is not in the document]
Comment (erikbern) [reply ann_benchmarks/runner.py:86]: i'm not following why this was rewritten this way to be honest... some more clarification would be helpful

**Comment by BossZou** [comment]
I'm very sorry that I pushed a bad request here. I'll close it.

**Comment by maumueller** [comment]
I actually think there were some really nice changes in there! I think it's fine to have it as a work in progress PR and modify it a bit?

**Comment by tinkerlin** [comment]
Most of changes in this PR are related to  the Milvus test, and we will pick a few useful commits to create a new PR. @maumueller

**Comment by erikbern** [comment]
Great, if there's any way to break this up in a few smaller PRs to be reviewed independently, that would be great!

**Comment by erikbern** [comment]
@BossZou any interest in resurrecting this? would love to see the newest version of Milvus integrated

**Pull request closed without merging**